zkir/Dialect/Field/Conversions/FieldToModArith/ExtensionFieldCodeGen.h

```diff
@@ -18,32 +18,106 @@ limitations under the License.
+#include "zkir/Dialect/Field/Conversions/FieldToModArith/FrobeniusCoeffs.h"
 #include "zkir/Dialect/Field/Conversions/FieldToModArith/PrimeFieldCodeGen.h"
+#include "zkir/Dialect/Field/Conversions/FieldToModArith/VandermondeMatrix.h"
 #include "zkir/Dialect/Field/IR/FieldTypes.h"
-#include "zkir/Dialect/ModArith/IR/ModArithOps.h"
 
 namespace mlir::zkir::field {
 
+// Forward declaration
+template <size_t N>
+class ExtensionFieldCodeGen;
+
+} // namespace mlir::zkir::field
+
+namespace zk_dtypes {
+
+template <size_t N>
+class ExtensionFieldOperationTraits<
+    mlir::zkir::field::ExtensionFieldCodeGen<N>> {
+public:
+  // TODO(chokobole): Support towers of extension field.
+  using BaseField = mlir::zkir::field::PrimeFieldCodeGen;
+  static constexpr size_t kDegree = N;
+};
+
+} // namespace zk_dtypes
+
+namespace mlir::zkir::field {
+
+// Selects the appropriate extension field operation based on degree.
+template <size_t N>
+struct ExtensionFieldOperationSelector;
+
+template <>
+struct ExtensionFieldOperationSelector<2> {
+  template <typename Derived>
+  using Type = zk_dtypes::QuadraticExtensionFieldOperation<Derived>;
+};
+
+template <>
+struct ExtensionFieldOperationSelector<3> {
+  template <typename Derived>
+  using Type = zk_dtypes::CubicExtensionFieldOperation<Derived>;
+};
+
+template <>
+struct ExtensionFieldOperationSelector<4> {
+  template <typename Derived>
+  using Type = zk_dtypes::QuarticExtensionFieldOperation<Derived>;
+};
+
 template <size_t N>
 class ExtensionFieldCodeGen
-    : public zk_dtypes::QuadraticExtensionFieldOperation<
-          ExtensionFieldCodeGen<N>> {
+    : public ExtensionFieldOperationSelector<N>::template Type<
+          ExtensionFieldCodeGen<N>>,
+      public VandermondeMatrix<ExtensionFieldCodeGen<N>>,
+      public FrobeniusCoeffs<ExtensionFieldCodeGen<N>, N> {
+  using Base = typename ExtensionFieldOperationSelector<N>::template Type<
+      ExtensionFieldCodeGen<N>>;
+
 public:
+  // Bring base class operator* into scope. Without this, the derived class's
+  // operator*(const PrimeFieldCodeGen&) hides all operator* overloads from
+  // Base.
+  using Base::operator*;
+  // TODO(junbeomlee): Remove these using declarations after refactoring
+  // zk_dtypes to not define these methods in QuarticExtensionFieldOperation.
+  using VandermondeMatrix<
+      ExtensionFieldCodeGen<N>>::GetVandermondeInverseMatrix;
+  using FrobeniusCoeffs<ExtensionFieldCodeGen<N>, N>::GetFrobeniusCoeffs;
+
   ExtensionFieldCodeGen(ImplicitLocOpBuilder *b, Type type, Value value,
                         Value nonResidue)
       : b(b), type(type), value(value), nonResidue(nonResidue) {}
   ~ExtensionFieldCodeGen() = default;
 
   operator Value() const { return value; }
 
-  // zk_dtypes::QuadraticExtensionFieldOperation methods
+  // Accessors for mixin classes
+  ImplicitLocOpBuilder *getBuilder() const {
+    assert(b);
+    return b;
+  }
```

Comment:
It's up to you, but we tend to use `ImplicitLocOpBuilder &` as a reference, so I suggest you to return it as reference instead of pointer.

```suggestion
  ImplicitLocOpBuilder &getBuilder() const {
    assert(b);
    return *b;
  }
```